async_server.c

```diff
@@ -0,0 +1,364 @@
+{
+  acl_context_t *cl_ctx = *(acl_context_t **)arg;
+  mt_context_t *mt_ctx = &cl_ctx->context->context;
+
+  pthread_cleanup_push (thread_cleanup_helper, cl_ctx);
+  while (true)
+    {
+      result_t task;
+      if (recv_wrapper (cl_ctx->socket_fd, &task, sizeof (task), 0)
+          == S_FAILURE)
+        {
+          error ("Could not receive result from client");
+          break;
+        }
+
+      if (task.is_correct)
+        {
+          if (queue_cancel (&mt_ctx->queue) == QS_FAILURE)
+            {
+              error ("Could not cancel a queue");
+              break;
+            }
+          memcpy (mt_ctx->password, task.password, sizeof (task.password));
+        }
+
+      trace ("Received %s result %s from client",
+             task.is_correct ? "correct" : "incorrect", task.password);
+
+      if (queue_push (&cl_ctx->registry_idx, &task.id) != QS_SUCCESS)
+        {
+          error ("Could not return id to a queue");
+          break;
+        }
+
+      trace ("Pushed index of received task back to indices queue");
+
+      cl_ctx->registry_used[task.id] = false;
+
+      if (serv_signal_if_found (mt_ctx) == S_FAILURE)
+        break;
+
+      if (mt_ctx->password[0] != 0)
+        break;
+    }
+
+  trace ("Cleaning up receiver thread");
+  pthread_cleanup_pop (!0);
+
+  return (NULL);
+}
+
+static void *
+task_sender (void *arg)
+{
+  acl_context_t *cl_ctx = *(acl_context_t **)arg;
+  mt_context_t *mt_ctx = &cl_ctx->context->context;
+
+  pthread_cleanup_push (thread_cleanup_helper, cl_ctx);
+  if (send_config_data (cl_ctx->socket_fd, mt_ctx) == S_FAILURE)
+    goto cleanup;
+
+  while (true)
+    {
+      int id;
+      if (queue_pop (&cl_ctx->registry_idx, &id) != QS_SUCCESS)
+        break;
+
+      trace ("Got index from registry");
+
+      task_t *task = &cl_ctx->registry[id];
+      if (queue_pop (&mt_ctx->queue, task) != QS_SUCCESS)
+        {
+          if (queue_push (&cl_ctx->registry_idx, &id) != QS_SUCCESS)
+            error ("Could not push back id to registry indices queue");
+          break;
+        }
+
+      trace ("Got task from global queue");
+
+      cl_ctx->registry_used[id] = true;
+
+      task_t task_copy = *task;
+      task_copy.task.id = id;
+      task_copy.to = task->from;
+      task_copy.from = 0;
+
+      if (send_task (cl_ctx->socket_fd, &task_copy) == S_FAILURE)
+        {
+          error ("Could not send task to client");
+          if (queue_push_back (&mt_ctx->queue, task) != QS_SUCCESS)
+            error ("Could not push back task to global queue");
+          cl_ctx->registry_used[task->task.id] = false;
+          break;
+        }
+    }
+
+cleanup:
+  trace ("Cleaning up sender thread");
+  pthread_cleanup_pop (!0);
+
+  return (NULL);
+}
+
+static void *
+handle_clients (void *arg)
+{
+  serv_context_t *srv_ctx = *(serv_context_t **)arg;
+  mt_context_t *mt_ctx = &srv_ctx->context;
+
+  while (true)
+    {
+      acl_context_t *acl_ctx = acl_context_init (srv_ctx);
```

Comment:
acl_context_destroy should be in cleanup handler for deallocation in case of thread cancelation

Reply:
Done

Reply:
Done